### PR TITLE
Make functions in Rewritable tailrec

### PR DESCRIPTION
**Pull request by davidegrohmann**

### CODE REVIEW

File: community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala [hunk 1]
```diff
@@ -119,53 +119,75 @@ object inSequence {
     }
   }
 
-  def apply(rewriters: Rewriter*) = new InSequenceRewriter(rewriters)
+  def apply(rewriters: Rewriter*): Rewriter = new InSequenceRewriter(rewriters)
```
Comment (Mats-SX) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala:122]: `override`

File: community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala [hunk 2]
```diff
@@ -119,53 +119,75 @@ object inSequence {
...
 }
 
 object topDown {
-
-  class TopDownRewriter(rewriter: Rewriter) extends Rewriter {
+  private class TopDownRewriter(rewriter: Rewriter) extends Rewriter {
     def apply(that: AnyRef): AnyRef = {
```
Comment (Mats-SX) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala]: `override`

File: community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala [hunk 3]
```diff
@@ -119,53 +119,75 @@ object inSequence {
...
+        val rewrittenJob = newJob.rewrite(rewriter)
+        stack.push((rewrittenJob :: jobs, doneJobs))
+        stack.push((rewrittenJob.children.toList, new mutable.MutableList()))
+        rec(stack)
+      }
     }
   }
 
-  def apply(rewriter: Rewriter) = new TopDownRewriter(rewriter)
+  def apply(rewriter: Rewriter): Rewriter = new TopDownRewriter(rewriter)
```
Comment (Mats-SX) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala]: `override` EDIT: No, wait this one is on the `object`, perhaps?

File: community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala [hunk 4]
```diff
@@ -119,53 +119,75 @@ object inSequence {
...
 }
 
 object bottomUp {
-
-  class BottomUpRewriter(val rewriter: Rewriter) extends Rewriter {
+  private class BottomUpRewriter(val rewriter: Rewriter) extends Rewriter {
     def apply(that: AnyRef): AnyRef = {
```
Comment (Mats-SX) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala]: `override`

File: community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala [hunk 5]
```diff
@@ -119,53 +119,75 @@ object inSequence {
...
+        }
+      } else {
+        stack.push((currentJobs.head.children.toList, new mutable.MutableList()))
+        rec(stack)
+      }
     }
   }
 
-  def apply(rewriter: Rewriter) = new BottomUpRewriter(rewriter)
+  def apply(rewriter: Rewriter): Rewriter = new BottomUpRewriter(rewriter)
```
Comment (Mats-SX) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala]: `override`

File: community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala (hunk 1 above)
Comment (davidegrohmann) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala:122]: On an object?

Comment (Mats-SX) [inline community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala:122]: You're right; I'm wrong.

**Pull request merged**